rfcs/20190305-modular-tensorflow.md

```diff
@@ -0,0 +1,481 @@
+*   Operating system (Linux, windows, macos, mobile)
+*   Architecture (x86, arm, ppc64, …)
+*   Accelerator (CPU, GPU,  TPU)
+*   Compiler (GCC, Clang, MSVC)
+*   Python version (2, 3.4, 3.5, 3.6,  3.7)
+
+More can be added to this list where we have to rebuild TensorFlow to support different network architectures, CUDA versions, SIMD instruction sets, etc.
+
+Having a monolithic repository means we need to rebuild all of our code for all of these different combinations. However, it makes no sense to rebuild all of our C++ code if the only difference is the Python version. Or rebuild all of our CPU kernels for different CUDA versions. Modularizing our code means we only need to rebuild and test the modules that are directly impacted by the dimensions we are changing in the support matrix.
+
+
+## Overview
+
+This project aims to split the TensorFlow codebase into **smaller, more focused**, repositories that can be released and managed separately. These modules will talk to each other using **well defined APIs** that guarantee backwards compatibility. Thanks to these APIs, these modules will be **managed/owned/released independently**. There will be different strategies to break apart pieces based on the languages, but below summarizes the approach for C++ and Python:
+
+
+![alt_text](20190305-modular-tensorflow/big_picture.png "Overview of modular TensorFlow")
+
+A summary of the above is:
+
+
+
+*   Core TF functionality will be implemented in C++
+*   Core TF functionality can be extended using shared objects.
+*   On top of the core C++ libraries, we will have the language bindings (Using the C API)
+*   There can be more functionality built on top of the core TF bindings in different languages, which can be maintained and distributed separately.
+*   All different pieces need to use Stable public APIs with backwards compatibility guarantees.
+
+
+### Definitions
+
+This section will briefly describe the terms we will use in the rest of this design.
+
+**Modules:** These are the components of core TF library that will "accept" plugins to expand their capabilities. Examples of modules are networking, filesystem, (graph) optimizers.
+
+**Plugins:** Plugins are extensions to different modules. For example, filesystem module can have a GCP plugin, an S3 plugin, an HDFS plugin.
+
+**Shared objects:** These are  dll/so/dylib files that can house **one or more** plugins for **one or more** modules.
+
+**Packages:** Python pip packages which may include Python files and/or shared objects.
+
+
+### C++
+
+This project aims to implement similar plugin architectures for multiple components of TF code. While these will be listed separately, **there will be a lot of shared pieces between these components**. The modules we would like to handle are:
+
+1.  Networking module, with verbs, gdr plugins initially
+1.  Filesystems module, with GCP, AWS and HDFS support
+1.  Kernels module, 
+1.  Optimizers/Graph rewrite module,
```

Comment:
(1) Will optimizers/graph rewrite modules offer same capabilities (such as different phases of rewrite such as pre-placement, post-placement, etc) that exists in current design? (2) What if plugins/backends want to have plugin-specific or backend-specific optimizers?

Reply:
Also, relevant question is if a plugin/backend needs a specific core component. To be precise, MKL-CPU backend currently uses BFCAllocator from TF core. How can we continue to ensure that this dependency is met?

Reply:
Our aim is to provide all the current capabilities, and we are working with grappler team and a few more groups internally to see how we should rework it. But our aim is to functionally match the current way the graph rewriting is possible with checking code into tensorflow.

Currently, for kernels our plan is to provide semantic equivalents to everything in OpKernelContext. While I cannot see a public interface to BFCAllocator there, we will be able to add more symbols to the API, as long as previous symbols are not changed.